### PR TITLE
Command Palette tweaks and docs

### PR DESCRIPTION
**Pull request by willmcgugan**
- Adds docs on the Command Palette to Guide.
- Change the module name from `command_palette` to `command`
- Changes the name of some the objects, mostly for brevity
- Adds a `post_init` function to the command sources.
- Makes textual._fuzzy public

### CODE REVIEW

File: src/textual/app.py
Comment (rodrigogiraoserrao) [inline src/textual/app.py:332]: We should use paths to textual objects because it's easier to keep URLs synced that way.
The `[some text here][textual.module.Object]` ones.

File: src/textual/command.py
Comment (rodrigogiraoserrao) [inline src/textual/command.py:4]:
```suggestion
```

File: src/textual/command.py
Comment (rodrigogiraoserrao) [inline src/textual/command.py:49]: Alphabetical order, please? 🥹

File: src/textual/command.py
Comment (rodrigogiraoserrao) [inline src/textual/command.py:63]:
```suggestion
    """The score of the command hit.
```

File: src/textual/command.py
Comment (rodrigogiraoserrao) [inline src/textual/command.py:111]:
```suggestion
    the method [`search`][textual.command.Source.search].
```

File: src/textual/command.py
Comment (rodrigogiraoserrao) [inline src/textual/command.py:153]:
```suggestion
            case_sensitive: Should matching be case sensitive?
```

File: src/textual/command.py
Comment (rodrigogiraoserrao) [inline src/textual/command.py:501]: Why are we making this public?

File: src/textual/screen.py
Comment (rodrigogiraoserrao) [inline src/textual/screen.py:161]: API link should be of the type `[text here][textual.module.Object]`

File: tests/snapshot_tests/snapshot_apps/command_palette.py
Comment (rodrigogiraoserrao) [inline tests/snapshot_tests/snapshot_apps/command_palette.py:2]:
```suggestion
from textual.command import Hit, Hits, Source
```

**Comment by rodrigogiraoserrao** [comment]
I left a bunch of comments and suggestions, but nothing blocking.

File: docs/examples/guide/command_palette/command01.py
Comment (davep) [reply docs/examples/guide/command_palette/command01.py:45]: The decision was made a wee while ago (so as part of the original PR) that it wouldn't be the case for the command sources; with the aim to review how people use this.

File: src/textual/command.py
Comment (davep) [inline src/textual/command.py:100]: Attributes of a `NamedTuple` are normally read-only. Is there something special about a `__post_init__` that allows this assignment?

**Review by davep — CHANGES_REQUESTED** [review]
From what I can see, the move to make the `Input`-bound text of the `Hit` optional isn't quite right and won't work; I've pointed to where I think the issue is.

I also want to test what happens if someone uses more interesting match text without supplying the `Input`-bound text.

**Comment by davep** [comment]
I did forget to add though: the docs are looking good!

File: src/textual/command.py
Comment (davep) [reply src/textual/command.py:100]: To see this using the example in the docs, add this following to `ViewerApp`:

```python
    def on_mount(self) -> None:
        from textual.command import CommandPalette
        CommandPalette.run_on_select = False
```

When you select a file the `Input` will read `None`.

File: docs/examples/guide/command_palette/command01.py
Comment (willmcgugan) [reply docs/examples/guide/command_palette/command01.py:45]: We might revisit that in the future. But I suspect the most common requirement will to override it per app, so inheritance is not that useful.

File: docs/guide/command_palette.md
Comment (willmcgugan) [reply docs/guide/command_palette.md:79]: The command sources are constructed each time the command palette is opened. I think `post_init` is somewhat familiar, as it has much the same meaning as dataclass `__post_init__`

File: src/textual/command.py
Comment (darrenburns) [reply src/textual/command.py:100]: __post_init__ only works on dataclasses

Comment (davep) [reply src/textual/command.py:100]: Also, to emphasise the part I mention about the intention of `match_display` vs `text`, it's worth keeping in mind that a `match_display` is a `RenderableType`. This should mean, and is done with the intention, that the display could be done within a Rich `Table` or some other more complex display. But the `str` of a `Table` would, unless I've missed something, be something like this:

```
<rich.table.Table object at 0x1022a6170>
```

If this is going to be an acceptable case of "leave off the `text` if you know your `match_display` can `str` to something readable and useable within an `Input`", we might want to tweak the documentation here to make it clear what the dev's responsibilities are regarding using the default case or not.

File: docs/guide/command_palette.md
Comment (willmcgugan) [reply docs/guide/command_palette.md:75]: I think I'd prefer to keep headers as text.

File: docs/guide/command_palette.md
Comment (willmcgugan) [reply docs/guide/command_palette.md:89]: The original reads better to me

Comment (davep) [reply docs/guide/command_palette.md:89]: I'll second that; I wouldn't personally put "of" in there.

File: src/textual/command.py
Comment (willmcgugan) [reply src/textual/command.py:4]: I don't think we should be too fussed about additional lines in docstrings.

File: src/textual/command.py
Comment (willmcgugan) [reply src/textual/command.py:111]: "Implement the method search" ? Maybe "implement the search method", but think it reads ok as is.

File: src/textual/screen.py
Comment (willmcgugan) [reply src/textual/screen.py:161]: I think Dave's intent was to link to the page in the API and not a specific object. How else would you do that?

File: src/textual/command.py
Comment (willmcgugan) [reply src/textual/command.py:501]: quells an error about overriding things. I don't think anyone will extend this anyway,

File: docs/examples/guide/command_palette/command01.py
Comment (willmcgugan) [reply docs/examples/guide/command_palette/command01.py:19]: I wouldn't have infered that. But how about "Called prior to searching" to avoid confusion?

File: docs/examples/guide/command_palette/command01.py
Comment (rodrigogiraoserrao) [reply docs/examples/guide/command_palette/command01.py:45]: Ok. I would've asked on the original PR but I was waiting for the green light to review the PR and it never came :P

File: src/textual/screen.py
Comment (davep) [reply src/textual/screen.py:161]: Sounds about right to me: if I recall correctly the intent was that it eventually settles on somewhere where the concept of command sources, what they are, how to use them, how to build them, etc; not on a specific bit of class documentation.

File: src/textual/command.py
Comment (rodrigogiraoserrao) [reply src/textual/command.py:4]: I wasn't going to make a PR to delete them, but this was one click away 🤷

File: docs/guide/command_palette.md
Comment (willmcgugan) [reply docs/guide/command_palette.md:9]: I don't know if we have settled on something else yet. We can change that in another PR.

File: src/textual/command.py
Comment (davep) [reply src/textual/command.py:501]: Nit-pick: if going public, we can lose the `_: Mount`.

File: docs/guide/command_palette.md
Comment (willmcgugan) [reply docs/guide/command_palette.md:13]: I think its fine as an opening paragraph. I don't like having to caveat everything.

File: docs/examples/guide/command_palette/command01.py
Comment (davep) [reply docs/examples/guide/command_palette/command01.py:19]: Perfect, I think that's nice and unambiguous!

File: docs/guide/command_palette.md
Comment (rodrigogiraoserrao) [reply docs/guide/command_palette.md:79]: It looks familiar but that familiarity is misleading: the `__post_init__` in dataclasses runs _once_ after instantiation.
Is the command source _instantiated_ every single time we open it?

If it is, then I retract my comment about `post_init` being a bad name and instead request that we make it _very_, _very_ clear that the command source is instantiated every time the palette is opened.

File: src/textual/command.py
Comment (willmcgugan) [reply src/textual/command.py:100]: Strange. I made `text` allow None, and assumed I would have to populate it. But this code will never even run, so I guess thats already taken care of elsewhere.

File: src/textual/screen.py
Comment (rodrigogiraoserrao) [reply src/textual/screen.py:161]: `/api/xxx` pages can be linked to with `[textual.xxx]`, e.g.:

```suggestion
    """The [command sources][textual.command] for the screen."""
```

If you don't want to switch to that link style, then at least the link requires fixing:

```suggestion
    """The [command sources](/api/command/) for the screen."""
```

File: docs/guide/command_palette.md
Comment (davep) [reply docs/guide/command_palette.md:13]: Cool, I can go with that; just as long as we're aware of it and are happy with it.

File: docs/guide/command_palette.md
Comment (davep) [reply docs/guide/command_palette.md:9]: Sure thing. But before release, yeah?

File: src/textual/command.py
Comment (davep) [reply src/textual/command.py:100]: As mentioned elsewhere; the upshot of this is that the `text` that's intended for the `Input` ends up being `"None"` -- this will be even more important if/when we do the thing we were talking about a while back, where a command can complete into the `Input` and then require further parameters. This is why the distinction between the `match_display` (which goes into the dropdown, and can be any renderable type: highlighted text, a Rich Table, probably a Rich Pixels Pokemon image, etc...) and the `text` is important here.

Comment (willmcgugan) [reply src/textual/command.py]: Ok, some misunderstandings on my part. I've made thing work the way I assumed they were! The `text` is populated automatically now, if `match_display` is a str or Text, which will be the 99% case.

Comment (davep) [reply src/textual/command.py]: Perfect! Rich Pixels-based Pokédex command source is safe once again!

File: docs/guide/command_palette.md
Comment (willmcgugan) [reply docs/guide/command_palette.md:9]: yup.

**Review by davep — APPROVED** [review]
Given my main concern here was the display text vs the input text... I'm all on board with this PR now.

File: docs/guide/command_palette.md
Comment (davep) [inline docs/guide/command_palette.md:106]:
```suggestion
This is useful for commands that only make sense when a given screen is active.
```

**Pull request merged**
